server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -777,13 +777,35 @@ public DiscoveryNodes nodes() {
             return nodes;
         }
 
+        // Deprecate to keep downstream projects compiling
+        @Deprecated(forRemoval = true)
         public Builder putTransportVersion(String nodeId, TransportVersion transportVersion) {
-            // TODO[wrb]: system index mappings versions will be added in a followup
-            compatibilityVersions.put(nodeId, new CompatibilityVersions(Objects.requireNonNull(transportVersion, nodeId), Map.of()));
+            return putCompatibilityVersions(nodeId, transportVersion, Map.of());
+        }
+
+        public Builder putCompatibilityVersions(
+            String nodeId,
+            TransportVersion transportVersion,
+            Map<String, SystemIndexDescriptor.MappingsVersion> systemIndexMappingsVersions
```

Comment:
Minor: it seems to me that this method is always called with an empty map. Maybe we could remove the parameter and give it a clearer name, like `putCompatibilityVersionsWithNoSystemIndexMapping` (I know, it's long, maybe there is something better). But something that makes the caller aware they are skipping something.

Reply:
The API I'm aiming for here is:
1. a method that puts a CompatibilityVersions object to the map, keyed to node ID
2. a method that sets the entire map of CompatibilityVersions for the builder
3. a convenience method that takes all the kinds of versions and constructs a CompatibilityVersions object. As we add new kinds of versions, we'll add arguments to this method.

Since "CompatibilityVersions" is already plural, it's a little awkward to name the method that takes the whole map. I think "nodeIdToCompatibityVersionsMap" might be a more informative name.

But I think I want to keep the three-arg compatibility versions method as-is. Eventually we'll add more things to CompatibilityVersions, and I don't want to add methods to handle subsets of available types of versions.